crates/profiling/src/jitdump.rs

```diff
@@ -0,0 +1,673 @@
+        size: u64,
+        file_suffix: Option<&str>,
+    ) -> Result<(), Error> {
+        let mut timestamp: u64 = 0;
+        JitDumpAgent::get_time_stamp(&mut timestamp);
+        if let Some(program) = unit.line_program.clone() {
+            let mut debug_info_record = DebugInfoRecord {
+                header: RecordHeader {
+                    id: RecordId::JitCodeDebugInfo as u32,
+                    record_size: 0,
+                    timestamp: timestamp,
+                },
+                address: address,
+                count: 0,
+            };
+
+            let mut debug_entries = Vec::new();
+            let mut debug_entries_total_filenames_len = 0;
+            let mut rows = program.rows();
+            while let Some((header, row)) = rows.next_row()? {
+                let row_file_index = row.file_index() - 1;
+                let myfile = dwarf
+                    .attr_string(
+                        &unit,
+                        header.file_names()[row_file_index as usize].path_name(),
+                    )
+                    .unwrap();
+                let filename = myfile.to_string_lossy()?;
+                let line = row.line().unwrap_or(0);
+                let column = match row.column() {
+                    gimli::ColumnType::Column(column) => column,
+                    gimli::ColumnType::LeftEdge => 0,
+                };
+
+                if (row.address() < address) || (row.address() > (address + size)) {
+                    continue;
+                }
+                let mut debug_entry = DebugEntry {
+                    address: row.address(),
+                    line: line as u32,
+                    discriminator: column as u32,
+                    filename: filename.to_string(),
+                };
+
+                if let Some(suffix) = file_suffix {
+                    debug_entry.filename.push_str("::");
+                    debug_entry.filename.push_str(suffix);
+                }
+
+                debug_entries_total_filenames_len += debug_entry.filename.len() + 1;
+                debug_entries.push(debug_entry);
+            }
+
+            debug_info_record.count = debug_entries.len() as u64;
+
+            let debug_entries_size = (debug_info_record.count
+                * (mem::size_of::<DebugEntry>() as u64 - mem::size_of::<String>() as u64))
+                + debug_entries_total_filenames_len as u64;
+            debug_info_record.header.record_size =
+                mem::size_of::<DebugInfoRecord>() as u32 + debug_entries_size as u32;
+
+            let _ = self.write_debug_info_record(debug_info_record);
+            let _ = self.write_debug_info_entries(debug_entries);
+        }
+        Ok(())
+    }
+}
+
+impl Clone for JitDumpAgent {
+    fn clone(&self) -> Self {
+        Self {
+            jitdump_file: self.jitdump_file.try_clone().unwrap(),
+            code_index: self.code_index,
+            dump_funcs: self.dump_funcs,
+        }
```

Comment:
I don't think `Clone` is needed for `JitDumpAgent`: it used just workaround mutability of `jitdump_file`. You can hide this detail using e.g. `RefCell`.